Updates to batch process

davidwaroquiers opened the pull request:
- Some refactoring (splitting up the big update_batch method)
- Renaming of variables
- Added possibility to have a batch collection to keep info about old batches (optional)

davidwaroquiers committed:
Modified batch part.
Renamed some variables (e.g. process_uuid to batch_uid, ...).
Splitted the batch method in the runner.
Miscellanous changes.

davidwaroquiers committed:
Merge branch 'develop' into dw/batch_process_id

davidwaroquiers committed:
Added missing patch_cli_consoles.

davidwaroquiers committed:
Added batches collection + miscellanous.

davidwaroquiers committed:
Fixed batch list.
Added some tests.

davidwaroquiers commented:
I've added one dependency (bidict) but I think I would remove it and deal with the "inverse" dict directly here

davidwaroquiers committed:
Merge branch 'develop' into dw/batch_process_id

davidwaroquiers committed:
Trying to fix test on CI.

davidwaroquiers committed:
Setting width for the patched consoles.

gpetretto commented on src/jobflow_remote/cli/batch.py:
```diff
@@ -30,6 +34,8 @@ def processes_list(
         ),
     ] = None,
+    show_all: show_all_batches_opt = False,
+    max_batches: max_batches_per_worker_opt = 10,
```
seems this option is never used. If to be implemented in the future I would suggest to comment it out for the moment.

davidwaroquiers replied:
Now it's done

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -129,6 +131,8 @@ def __init__(
             The name of the collection used to store other auxiliary data.
             Uses the DB defined in the queue_store.
+        batches_collection
+            The name of the collection used to store the batch processes.
```
Maybe also here it can be written that it is optional?

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4755,12 +4786,35 @@ def add_batch_process(
+                    "batch_state": BatchState.SUBMITTED.value,
+                    "worker": worker,
+                }
+            )
```
It is probably better like this, given the simple format of the document, but I was wondering if you considered creating a `BatchDoc` pydantic model to insert/retrieve the data. 

Below the `jobs` value is set in this dictionary. Maybe it can already be added a `"jobs": {}`? Or is it not worth?

Also, maybe it would be useful to add more information?
For example also here an `updated_on`? And also maybe `start_time` and `end_time`? So that one can understand if maybe a batch jobs are being submitted but without doing anything?

davidwaroquiers replied:
For the BatchDoc pydantic model, I indeed wondered if it was worth it. What do you think ?
I've set jobs as a dict from the beginning.
For the additional information, I've added updated_on, start_time and end_time. For start and end though I am not sure. This is the time at which the runner "finds out" that the batch process started or finished. It might be delayed (or even not found out in the case of start_time, if it directly goes to finished state).

gpetretto replied:
with this amount of information maybe it is worth to have a `BatchDoc` instead of working with dictionaries? Will such an object be used only for the CLI? Or will it give a benefit in other operations?

`start_time` and `end_time` should not be too difficult to get correctly, but can also be avoided as far as I am concerned.

davidwaroquiers replied:
Ok, I'll add the BatchDoc. It can be useful anyway.
I've added start and end time

davidwaroquiers replied:
BatchDoc Done

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4755,12 +4786,35 @@ def add_batch_process(
+            self.batches.update_one(
+                {"batch_uid": batch_uid, "worker": worker},
+                {"$set": {f"jobs.{job_id}.{job_index}": info}},
+            )
```
Should we consider a safeguard with respect to the maximum 16MB document size in case a lot of jobs are added? (not sure how)

davidwaroquiers replied:
Maybe just a try except that only does the update of the updated_on in case adding the job makes it go beyond the 16 Mb ? I would guess it would be an edge case no ? (would really need a lot of jobs, except if we start to add a lot of info ..)

gpetretto replied:
ok, let's leave it like this for the moment.

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4755,12 +4786,35 @@ def add_batch_process(
             upsert=True,
         )
 
+    def add_job_to_batch(self, job_id, job_index, batch_uid, worker, info):
```
missing typing.

Also the method name is confusing: I have seent that it is called multiple times and it seems it is done to update the current state of the Job, but initially I was confused by the fact that the same job was "added" multiple times.

davidwaroquiers replied:
Added typing and changed name to "update_job_in_batch"

gpetretto commented on src/jobflow_remote/cli/batch.py:
```diff
@@ -39,29 +45,77 @@ def processes_list(
+                }
+                for worker, worker_batches in batch_processes.items()
+                for process_id, batch_uid in worker_batches.items()
+            ]
```
The fact that here `batch_processes` gets reassigned with a different data structure is a bit confusing. Maybe the original value can be called something like `batch_processes_data` so that it is clear that they are two different structures?

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+                        job_index=job_index,
+                        batch_uid=batch_uid,
+                        worker=worker_name,
+                        info={"state": next_state.value},
```
it seems this is the last time the info is modified. Maybe it can always be TERMINATED? For a generic user it will not make much sense that in some cases it will be DOWNLOADED, in other TERMINATED, but it will never move further from that.

davidwaroquiers replied:
Should we update the batch documents with the real job states when they go beyond TERMINATED ? That would be an additional call to the database. Not sure if it's worth ... maybe manually if something goes wrong ?

gpetretto replied:
I would say it is not worth (and likely involved). For the batch process it is enough to know that the job was terminated. If one want more info should go to the job details.

davidwaroquiers replied:
Ok, I'll put TERMINATED (will even make more "sense" when it's named RUN_FINISHED)

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/cli/batch.py:
```diff
@@ -39,29 +45,77 @@ def processes_list(
+                    for jid, jid_dict in batch.get("jobs", {}).items()
+                    for jidx in jid_dict
+                ]
+                batches_jobs.append(batch_jobs)
```
It seems that this will not only include the running jobs, but all the jobs that were also terminated. Am I right? But in the table they are still indicated as "Running Job ids".

davidwaroquiers replied:
Correct, I changed

gpetretto replied:
if these are not just running jobs anymore maybe the name of the argument of `get_batch_processes_table` can also be changed, instead of `running_jobs`?

davidwaroquiers replied:
Yes, that makes sense. Done

gpetretto commented on src/jobflow_remote/cli/formatting.py:
```diff
@@ -586,39 +586,37 @@ def format_upgrade_actions(actions: list[UpgradeAction]):
+            get_job_path(batch_data["batch_uid"], None, worker.batch.work_dir),
+        ]
+        if status:
+            row.append(batch_data.get("batch_state", "UNKNOWN"))
```
Does "UNKNOWN" get printed for all the jobs if `show_all` is set to False? In that case I would suggest to set `status` to False as well. The "UNKNOWN" gives the feeling that there may be something wrong ("why is it unknown?"). While it is just that there is no way of knowing for the code and they will always be all unknown in that case.

davidwaroquiers replied:
Here it's the batch_state. If show_all is set to False, then it's just the jobflow jobs that are not shown, but the batch state can still be there. But I agree that, as a user, you can wonder why it's unknown. I don't know what to do here. Maybe put nothing if the batch_state is unknown ? row.append(batch_data.get("batch_state", ""))

gpetretto replied:
dosn't `show_all` make the difference between fetching from the auxiliary collection and from the batches one? So it seems that if show_all is False the batch state con only be `UNKNOWN`? 
An empty box will be better, but if the auxiliary is dropped this case will not happen anymore, right?

davidwaroquiers replied:
Ok I see. In the case of show_all = False, get_batch_processes_table is called with the default value of status (= to False). So I think it's already the case.
Indeed if we drop the auxiliary for batches it should not happen. I propose to resolve this conversation if what I'm saying is correct but maybe I misunderstood.
For dropping the auxiliary, let's see.

gpetretto replied:
I see, but so the case where `UNKNOWN` is used should never happen, right?

davidwaroquiers replied:
I think it should not indeed. You'd just get access to the key and it will then fail if the key does not exist (which again, should not happen) ?

davidwaroquiers replied:
I changed to batch_data["batch_state"] as it should never happen (if it happens, it's a bug).

gpetretto commented on src/jobflow_remote/jobs/batch.py:
```diff
@@ -53,6 +55,7 @@ def __init__(
         self.running_dir = self.files_dir / RUNNING_DIR
         self.terminated_dir = self.files_dir / TERMINATED_DIR
         self.lock_dir = self.files_dir / LOCK_DIR
+        self.batches_history_dir = self.files_dir / BATCHES_HISTORY_DIR
```
Is this used somewhere?

davidwaroquiers replied:
No, leftover. I removed everything related

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+        while len(self._cached_batches[worker_name]) > 2 * worker.max_jobs + 5:
+            self._cached_batches[worker_name].popitem(last=False)
+
+    def get_process_id(self, batch_processes, batch_uid, worker_name, worker):
```
maybe it can be renamed `batch_get_process_id` or `get_batch_process_id`?

missing typing

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+        return "UNKNOWN"
+
+    def batch_update_status(
+        self, batch_manager, queue_manager, worker_name, worker, running_jobs
```
missing typing

davidwaroquiers replied:
Added

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+            return list(running_processes)
+        return []
+
+    def submit_batch_processes(self, queue_manager, worker_name, worker, processes):
```
missing typing

davidwaroquiers replied:
Added

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
-                            }
+
+            for pid in running_processes:
+                self.job_controller.set_running_batch_process(pid, worker_name)
```
This keeps updating the state of the state of the running process for all the jobs, for all their duration in the DB, right? I suppose it is not very intensive action, but they are still a lot of calls to the DB for nothing.

davidwaroquiers replied:
True, and also the start_time and updated_on time is now updated every time ...
I have updated the set_running_batch_process to set it only when start_time is None.
At least it's correct now.
Still have to think about how to not do the db calls everytime (maybe caching also in the _cached_batches somehow ?)

gpetretto replied:
yes, I was thinking of keeping track of that in the runner. Of course it will try again if the runner is restarted, but at least only once.

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+            return doc["process_id"]
+        # What do we do when we don't have it ? (I'm pretty sure it cannot happen if we have the batches collection,
+        # but if we don't, it could happen, although relatively "unlikely")
+        return "UNKNOWN"
```
Is there a point in returning "UNKNOWN"? I think dealing with hardcoded string values is often a pain. Isn't there any better solution here? Would it be better to raise, since it should never happen?
Or what about returning `None`?

davidwaroquiers replied:
Yes, I just did not know what to do at the time, hence the comment :) I've put a RuntimeError. What do you think ?

gpetretto replied:
raise is fine with me

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+            # Check the batch processes that should be running on the remote queue
+            # and update the state in the DB if something changed (both for batch processes and
+            # for the states of the jobs that were running in these batch processes)
+            processes = self.batch_update_status(
```
Maybe the `processes` names can be more explicit? Like `running_processes`?

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+                    f"submission succeeded but ID not known. Job may be running but status cannot be checked. {submit_result!r}"
+                )
+                # TODO: should there be anything here to check with the uuid somehow on the worker ?
+                #  More specifically, maybe we could just add the batch process to the batches collection with None as an ID ?
```
quick comment on this point: if you add the entry with None as process_id, what if there are more than one? How do you tell them apart?

davidwaroquiers replied:
Fully agree ... no clue. Maybe from the name of the job ? Maybe there could be a workaround to get something from the remote worker in the batch handle directory ?

gpetretto replied:
honestly this seems a challenging point. Maybe we can leave it in case this really starts to happen?

davidwaroquiers replied:
ok I resolve

gpetretto commented on src/jobflow_remote/remote/queue.py:
```diff
@@ -224,7 +224,11 @@ def get_jobs_list(
+        # New versions of qtoolkit return the command and the list of stringified job_ids.
+        # This ensures backward compatibility.
+        result = self.scheduler_io.get_jobs_list_cmd(jobs, user)
+        job_cmd = result[0] if isinstance(result, tuple) else result
```
Can we first settle if it is really needed to change the API of QTK?

davidwaroquiers replied:
I will go back to qtk to see about that. I agree

davidwaroquiers replied:
Removed for now

gpetretto commented on tests/conftest.py:
```diff
@@ -538,15 +543,28 @@ def update_project_data(
+        # case where patch_project is not parametrized but just used inside the test
+        yield partial(
+            update_project_data, project_file_path=Path(current_project_data.filepath)
+        )
```
Maybe it would be worth making two different fixtures for the two different cases? One might be tempted to parametrize **and** also try to call `patch_project` to make some further tests during the test execution. Since in this case it does not yield nothing, this will result in an error.

Or maybe always yield `partial(update_project_data, project_file_path=Path(current_project_data.filepath))`?

davidwaroquiers replied:
I did it with your second proposal (always yield the partial)

gpetretto commented on tests/integration/test_advanced_options.py:
```diff
@@ -29,21 +41,50 @@ def test_run_batch(job_controller, monkeypatch, clean_slurm_queue) -> None:
+    monkeypatch.setattr(runner.runner_options, "delay_update_batch", 0.05)
+    monkeypatch.setattr(runner.runner_options, "delay_advance_status", 0.05)
+    monkeypatch.setattr(runner.runner_options, "delay_check_run_status", 0.05)
+    monkeypatch.setattr(runner.runner_options, "delay_checkout", 0.05)
```
Should we consider doing this generally during the tests? Does it really speed things up with such small timings?

davidwaroquiers replied:
It does speed up the tests indeed. I've also made the changes to the config base so that all timings can be expressed in floats. I guess we can see in another PR to speed up tests that are long

gpetretto replied:
ok, let's keep it for another PR.

gpetretto commented on tests/integration/test_advanced_options.py:
```diff
@@ -29,21 +41,50 @@ def test_run_batch(job_controller, monkeypatch, clean_slurm_queue) -> None:
+    elif request.node.callspec.id == "without_batches_collection":
+        assert runner.job_controller.batches is None
+    else:
+        pytest.fail("Should not be here!")
```
The message here maybe can be a bit informative? Like "Wrong parametrization id: {request.node.callspec.id}"?

davidwaroquiers replied:
Done

gpetretto commented on tests/integration/test_advanced_options.py:
```diff
@@ -29,21 +41,50 @@ def test_run_batch(job_controller, monkeypatch, clean_slurm_queue) -> None:
     # This should test that the batch runner is not running with multiple
     # parallel processes
-    jobs_info = job_controller.get_jobs_info()
+    jobs_info = runner.job_controller.get_jobs_info()
```
I am curious, why is this needed?

davidwaroquiers replied:
This was before I modified the patch_project to be applied directly before the job_controller fixture. Now it can be removed.

gpetretto commented on tests/integration/test_advanced_options.py:
```diff
@@ -186,7 +280,23 @@ def submit_jobs(n: int, sleep: int):
+            required_out=["Batches info", "Status"],
+        )
+
+    assert len(batch_manager.get_running()) == 0
```
there is `get_running` twice

davidwaroquiers replied:
Removed

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1122,343 @@ def update_batch_jobs(self) -> None:
+            if batch_uid in worker_batches:
+                return worker_batches[batch_uid]
+        # Then try to take it from the running batch processes
+        process_id = batch_processes.inverse.get(batch_uid)
```
if the fact of being a bidict is used only here, maybe indeed the dependecy on bidict can be removed and just invert the dictionary on the fly if needed?
In general it seems that this bidict would be useful if the same dictionary needs to be updated and keep track of changes in both direction, but if I am not wrong here it is just instanciated and used as it is.

davidwaroquiers replied:
I agree. Removed.

gpetretto reviewed (commented):
Thanks for the updates.
I have left some comments.
A general one is that it seems that the batches collection could in principle replace entirely the auxiliary collection, am I right? Should we consider switching to that, instead of keeping duplicated information in the DB? Also, if someone is interested in using the batch, why would they not be interested in having all the possible information? Assuming that this is not too big.
This would be of course a backward incompatible change, so it might be a bit annoying to handle.

davidwaroquiers committed:
Added max-batches-per-worker implementation.
Added tests for batch worker locally.
Added tests for the max batch per worker option.

davidwaroquiers committed:
Fixed typing (linting) for timings to allow float values.

davidwaroquiers committed:
Modifications following code review.

davidwaroquiers committed:
Merge branch 'develop' into dw/batch_process_id

davidwaroquiers reviewed (commented):
I've made some updates. You can have a look and resolve conversations that are now fixed and we can discuss the remaining ones.

gpetretto commented on src/jobflow_remote/cli/batch.py:
```diff
@@ -39,29 +46,94 @@ def processes_list(
+            sort={"end_time": -1},
+        )
+
+        batch_processes.extend(finished_batch_processes)
```
It seems a bit weird to ask for maximum 10 jobs and still get more than 10. Why this?
Wouldn't it be easier to add the option to filter based on `BatchState` instead, so the user can decide?

davidwaroquiers replied:
Maybe we can do this then. We get max 10 jobs in total, but if there are more jobs, we inform how many more, e.g. if you were to have 15 batch processes running for one worker, you get 10 that are shown and a message saying something like "10/15 running batches shown"

davidwaroquiers replied:
This is now done with the max_results. Resolving.

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4731,9 +4749,45 @@ def get_batch_processes(
+                query["batch_state"] = batch_state.value
+            else:
+                query["batch_state"] = {"$in": [bs.value for bs in batch_state]}
+        sort = sort or {"created_on": -1}
```
for the jobs and flows by default we sort on `updated_on`. Don't know if we want to keep the same convention?

davidwaroquiers replied:
Yes, done

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4731,9 +4749,45 @@ def get_batch_processes(
+                },
+                {"$match": {"rank": {"$lte": max_batches_per_worker}}},
+                {"$unset": "rank"},
+            ]
```
I wonder if instead of making this complicated query, that I suppose may become more of a burden as the size of the collection grows, and is not really optimal for the user it wouldn't be easier to just have the options `--worker` and `--max-results`, so the user decides what to search for. Most likely one has only one batch worker, and if there are many it is not obvious that you want to see the last 10 jobs for every of these workers.
What you get would be clearer, more flexibility and the query much simpler, without the need to make an aggregation.

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4731,9 +4749,45 @@ def get_batch_processes(
+        batch_state: BatchState | list[BatchState] | None = None,
+        max_batches_per_worker=None,
+        sort=None,
+    ):
```
return type is missing

davidwaroquiers replied:
Done

gpetretto replied:
is it? It seems it does not show

davidwaroquiers replied:
Ah probably did not go through. Now done :)

gpetretto commented on src/jobflow_remote/jobs/runner.py:
```diff
@@ -1118,221 +1121,370 @@ def update_batch_jobs(self) -> None:
+        batch_uid: str,
+        worker_name: str,
+        worker: WorkerBase,
+    ):
```
return type missing

davidwaroquiers replied:
Done

gpetretto reviewed (commented):
Thanks for the updates. Here are some replies and additional small comments.
I suppose some changes can also be hold, depending on whether we drop the auxiliary document or not.

davidwaroquiers committed:
Added start and end times for batch processes.

davidwaroquiers committed:
Merge branch 'develop' into dw/batch_process_id

davidwaroquiers committed:
Added BatchDoc + other modifications.

davidwaroquiers committed:
Modifications of jf batch list.

davidwaroquiers committed:
Fixed test.

davidwaroquiers reviewed (commented):
More updates.
I think last thing is to decide to remove the dealing of the batches from the auxiliary collection (or do that later or not do that).
If you can go through the comments again and check if everything is solved and we can have a quick call to see what we do about the above.

davidwaroquiers committed:
Fixed linting.

gpetretto commented on src/jobflow_remote/cli/batch.py:
```diff
@@ -39,29 +37,87 @@ def processes_list(
+                    ]
+                    for batch in batch_processes
+                ]
+            table = get_batch_processes_table(
```
it seems that here the `max_results` is ignored. However it is unlikely that one will have more than 20 batch jobs running at the same time. And in addition the auxiliary collection could be removed. 
Is it fine to keep it like this?

davidwaroquiers replied:
I would indeed keep it like that and see if we switch to the no auxiliary collection for the batches

gpetretto commented on src/jobflow_remote/jobs/batch.py:
```diff
@@ -176,6 +178,15 @@ def cleanup(self) -> bool:
 
+    def get_batch_info(
+        self, batch_dir: Path | str, batch_info_file: str = BATCH_INFO_FILENAME
+    ):
```
missing return type

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/jobcontroller.py:
```diff
@@ -4731,9 +4750,32 @@ def get_batch_processes(
+                query["batch_state"] = {"$in": [bs.value for bs in batch_state]}
+        sort = sort or {"updated_on": -1}
+
+        return list(self.batches.find(query, sort=sort).limit(max_results))
```
I was expecting that this would return a BatchDoc, otherwise I see little point in adding the model if we don't use it to pass the data around. It is at this point used only to insert the data, right?

davidwaroquiers replied:
True, I thought that I had to do that then I forgot :) Will do!

davidwaroquiers replied:
Done

gpetretto commented on src/jobflow_remote/jobs/run.py:
```diff
@@ -119,81 +125,119 @@ def run_remote_job(run_dir: str | Path = ".") -> None:
+        time.sleep(interval)
+
+
+PING_TIME = 60
```
Pinging every minute seems an unnecessarily high frequency.

davidwaroquiers replied:
600 ? 3600 ?

davidwaroquiers replied:
I've put 600

davidwaroquiers committed:
Fixed bug for worker in get_all_batches.

gpetretto commented on src/jobflow_remote/jobs/run.py:
```diff
@@ -119,81 +125,119 @@ def run_remote_job(run_dir: str | Path = ".") -> None:
     max_jobs: int | None = None,
     parallel_jobs: int | None = None,
+    sleep_time: float = None,
+    batch_dir: str | Path = None,
```
I wonder if it makes sense that this is an option. If it is not in the CWD the runner will not even find it, since it assume it is there.

davidwaroquiers replied:
Removed

gpetretto reviewed (commented):
Thanks for the updates. I left a few more comments.

davidwaroquiers committed:
BatchDoc validation + various updates following review comments.

davidwaroquiers reviewed (commented):
I should have addressed all the comments.

gpetretto reviewed (approved):
LGTM. Thanks!
